environment.cpp

```diff
@@ -78,25 +78,52 @@ void Environment::clearMutagen(){
 
 void Environment::tick(const int amount /* = 1 */){
   for (int i = 0; i < amount; ++i) {
+    // Dots with energy <= 0 die
+    for(size_t i = 0; i < dots_vec_.size(); ++i){
+      if(dots_vec_[i].getEnergy() <= 0){
+        alive_vec_[i] = false;
+      }
+    }
     // Generate new Food
     feeding();
     // Trigger Tick of every Dot and let Dots with enough energy replicate
     searchFood();
+    cannibalism();
     for (size_t i = 0; i < dots_vec_.size(); ++i) {
+      if(!alive_vec_[i]){
+        continue;
+      }
       dots_vec_[i].tick();
-      if(dots_vec_[i].getReproductionCooldown() <= 0 && dots_vec_[i].getEnergy() >= 5000){
+      if(dots_vec_[i].getReproductionCooldown() <= 0 && dots_vec_[i].getEnergy() >= 5000 + 10 * dots_vec_[i].getSize()){
         double mutation_rate = 0.2; // the background_mutationrate is constant at the moment, change here if wanted
         for (size_t i = 0; i < mutagen_vec_.size(); ++i) {
           uint16_t distance_to_mutagen = distance(dots_vec_[i].getPosition(), mutagen_vec_[i], testground_size_);
           mutation_rate += 1/(distance_to_mutagen * distance_to_mutagen + 1);
         }
         dots_vec_.push_back(dots_vec_[i].replicate(mutation_rate));
+        alive_vec_.push_back(true);
+      }
+    }
+    // dots_vec_.erase(remove_if(dots_vec_.begin(), dots_vec_.end(), [](Dot &d){
+    //   return !(d.getAlive());
+    // }), dots_vec_.end());
+    // Reduce Indices of Preys
+    for(size_t i = 0; i < dots_vec_.size(); ++i){
+      if (dots_vec_[i].getPreyInSight()) {
+        if(alive_vec_[dots_vec_[i].getPreyInSightIdx()]){
+          dots_vec_[i].setPreyInSightIdx(dots_vec_[i].getPreyInSightIdx() - std::count(alive_vec_.begin(), alive_vec_.begin() + dots_vec_[i].getPreyInSightIdx(), false));
```

Comment:
std::count gets called for every hunter. It might probably be better if we store results in between, e.g.: If we compute count from begin to 100 first, we only need count from 100 to 110 for the second and add it to the first count.